agent/agent.go

```diff
@@ -168,52 +153,79 @@ func (r *Agent) Run(recipe recipe.Recipe) (run Run) {
 	return
 }
 
-// runExtractor runs an extractor.
-func (r *Agent) runExtractor(ctx context.Context, sourceRecipe recipe.SourceRecipe, in chan<- interface{}) (err error) {
-	extractor, err := r.extractorFactory.Get(sourceRecipe.Type)
+func (r *Agent) setupExtractor(ctx context.Context, sr recipe.SourceRecipe, str *stream) (runFn func() error, err error) {
+	extractor, err := r.extractorFactory.Get(sr.Type)
 	if err != nil {
+		err = errors.Wrapf(err, "could not find extractor \"%s\"", sr.Type)
 		return
 	}
-	if err = extractor.Extract(ctx, sourceRecipe.Config, in); err != nil {
+	err = extractor.Init(ctx, sr.Config)
+	if err != nil {
+		err = errors.Wrapf(err, "could not initiate extractor \"%s\"", sr.Type)
 		return
 	}
 
+	runFn = func() (err error) {
+		err = extractor.Extract(ctx, str.push)
+		if err != nil {
+			err = errors.Wrapf(err, "error running extractor \"%s\"", sr.Type)
```

Comment:
We also need to cancel this run if extractor fails I guess? This should not be a soft fail?

Reply:
Extractor and Processor errors should be hard fail I think, only sinks can soft fail.

Reply:
> We also need to cancel this run if extractor fails I guess? This should not be a soft fail?

this error will be escalated to https://github.com/odpf/meteor/pull/213/files?file-filters%5B%5D=.go#diff-62b6ad581fe3a3059ae8c85ef0f31dde4092bfdecfa7d6857c470bcacaa8cc8bR129

and it will close the stream immediately

Reply:
> Extractor and Processor errors should be hard fail I think, only sinks can soft fail.

right now everything is hard fail, but having sinks soft fail would be beneficial in multiple sinks case 🤔 .

will definitely consider that changes, but probably outside of this scope for now.

Reply:
Got it, missed that `stream.Close()` while handling error. Then this should be fine.